cpp/src/arrow/compute/kernels/isin.cc

```diff
@@ -0,0 +1,325 @@
+    }
+    return Status::OK();
+  }
+
+  // \brief Iterate over the left array using another visitor.
+  // In VisitValue, use the memo_table_ (for right array) and check if value
+  // in left array is in the memo_table_. Append the true to the
+  // BooleanBuilder if condition satisfied, else false.
+  Status VisitValue(const Scalar& value) {
+    bool_builder_.UnsafeAppend(memo_table_->Get(value) != -1);
+    return Status::OK();
+  }
+
+  Status Reset() override {
+    bool_builder_.Reset();
+    return Status::OK();
+  }
+
+  Status Compute(FunctionContext* ctx, const Datum& left, Datum* out) override {
+    const ArrayData& left_data = *left.array();
+    RETURN_NOT_OK(bool_builder_.Reserve(left_data.length));
+    RETURN_NOT_OK(ArrayDataVisitor<Type>::Visit(left_data, this));
+    RETURN_NOT_OK(bool_builder_.FinishInternal(&output));
+    out->value = std::move(output);
+    return Status::OK();
+  }
+
+  Status ConstructRight(FunctionContext* ctx, const Datum& right) override {
+    MemoTableRight<Type, Scalar> func;
+    RETURN_NOT_OK(func.Reset());
+
+    if (right.kind() == Datum::ARRAY) {
+      RETURN_NOT_OK(func.Append(ctx, right));
+    } else if (right.kind() == Datum::CHUNKED_ARRAY) {
+      const ChunkedArray& right_array = *right.chunked_array();
+      for (int i = 0; i < right_array.num_chunks(); i++) {
+        RETURN_NOT_OK(func.Append(ctx, right_array.chunk(i)));
+      }
+    } else {
+      return Status::Invalid("Input Datum was not array-like");
+    }
+
+    memo_table_ = std::move(func.memo_table_);
+    right_null_count = func.right_null_count;
+    return Status::OK();
+  }
+
+ protected:
+  using MemoTable = typename HashTraits<Type>::MemoTableType;
+  std::unique_ptr<MemoTable> memo_table_;
+  std::shared_ptr<DataType> type_;
+  MemoryPool* pool_;
+
+ private:
+  // \brief Additional member "right_null_count" is used to check if
+  // null count in right is not 0
+  int64_t right_null_count{};
+  BooleanBuilder bool_builder_;
+  std::shared_ptr<ArrayData> output;
+};
+
+// ----------------------------------------------------------------------
+// (NullType has a separate implementation)
+
+class NullIsInKernel : public IsInKernelImpl {
+ public:
+  NullIsInKernel(const std::shared_ptr<DataType>& type, MemoryPool* pool) {}
+
+  Status Reset() override {
+    bool_builder_.Reset();
+    return Status::OK();
+  }
+
+  // \brief When array is NullType, based on the null count for the arrays,
+  // append true to the BooleanBuilder else propagate to all nulls
+  Status Compute(FunctionContext* ctx, const Datum& left, Datum* out) override {
+    const ArrayData& left_data = *left.array();
+    left_null_count = left_data.GetNullCount();
+
+    if (left_null_count != 0 && right_null_count == 0) {
+      output = out->array();
+      output->type = boolean();
+      RETURN_NOT_OK(detail::PropagateNulls(ctx, left_data, output.get()));
+    } else {
+      RETURN_NOT_OK(bool_builder_.Reserve(left_data.length));
```

Comment:
see comment above,  we shouldn't be allocating in kernel when PrimiteAllocatingKernel is the preferred method to get an already allocated value.